src/core/storage/sframe_data/groupby_aggregate_impl.cpp

```diff
@@ -347,46 +395,140 @@ void group_aggregate_container::flush_segment(size_t segmentid) {
     delete hash_entries.second;
   }
 
-  for (auto& item: local_sorted) {
-    for (auto& value: item.values) {
+  for (auto& item : local_sorted) {
+    for (auto& value : item.values) {
       value->partial_finalize();
     }
   }
   // ok. now we can write! lock the file
-  std::unique_lock<turi::mutex> filelock(segments[segmentid].file_lock);
-  oarchive oarc;
-  for (auto& item: local_sorted) {
-    oarc << item;
-    // write into the iterator
-    *(segments[segmentid].outiter) = std::string(oarc.buf, oarc.off);
-    ++(segments[segmentid].outiter);
-    oarc.off = 0;
+  size_t round_robin = (++merry_go_round_) % local_buffer_set_.size();
+
+  auto& local_buffer = local_buffer_set_[round_robin];
+
+  local_buffer.refctr_++;
+  /// lock the segmeent from local_buffer before using it
+  {
+    std::lock_guard<turi::simple_spinlock> slk(
+        local_buffer.sa_seg_locks_[segmentid]);
+
+    logstream(LOG_INFO) << "flush buffer from task_id: " << tss_.id_
+                            << ", segment_id: "<< segmentid << ", on buffer: " << round_robin << std::endl;
+
+    auto outiter = local_buffer.sa_buffer_ptr_->get_output_iterator(segmentid);
+    oarchive oarc;
+    for (auto& item : local_sorted) {
+      oarc << item;
+      // write into the iterator
+      *(outiter) = std::string(oarc.buf, oarc.off);
+      ++(outiter);
+      oarc.off = 0;
+    }
+
+    free(oarc.buf);
+    local_buffer.sa_seg_chunks_[segmentid].push_back(local_sorted.size());
   }
-  free(oarc.buf);
-  segments[segmentid].chunk_size.push_back(local_sorted.size());
+  local_buffer.refctr_--;
+}
+
+void group_aggregate_container::merge_local_buffer_set() {
+  if (gl_buffer_.is_opened_for_write())
+    log_and_throw("intermediate_buffer is should be closed before stealing local_buffer.");
```

Comment:
This is an internal check, which can be properly handled with an ASSERT statement. Just do ASSERT_MSG(!gl_buffer_.is_opened_for_write(), "intermediate_buffer is should be closed before stealing local_buffer.");

Reply:
I will use this form in the future!